File: ui/src/lib/bff/hypervisor-settings.ts
import { bffFetch } from './client';
import { BFFEndpoints } from './endpoints';

export type HypervisorSettings = {
	cpu_nested: boolean;
	cpu_amx: boolean;
	cpu_kvm_hyperv: boolean;
	memory_mergeable: boolean;
	memory_hugepages: boolean;
	memory_shared: boolean;
	memory_prefault: boolean;
	iommu: boolean;
	rng_src: string;
	watchdog: boolean;
	landlock_enable: boolean;
	serial_mode: string;
	console_mode: string;
	pvpanic: boolean;
	tpm_type: string | null;
	tpm_socket_path: string | null;
	profile_id: string | null;
};

export type HypervisorProfile = {
	id: string;
	name: string;
	description: string | null;
	cpu_nested: boolean | null;
	cpu_amx: boolean | null;
	cpu_kvm_hyperv: boolean | null;
	memory_mergeable: boolean | null;
	memory_hugepages: boolean | null;
	memory_shared: boolean | null;
	memory_prefault: boolean | null;
	iommu: boolean | null;
	rng_src: string | null;
	watchdog: boolean | null;
	landlock_enable: boolean | null;
	serial_mode: string | null;
	console_mode: string | null;
	pvpanic: boolean | null;
	tpm_type: string | null;
	tpm_socket_path: string | null;
	is_builtin: boolean;
};

export type HypervisorSettingsResponse = {
	settings: HypervisorSettings;
	profiles: HypervisorProfile[];
};

export type HypervisorSettingsPatch = Partial<HypervisorSettings>;

export async function getHypervisorSettings(token?: string): Promise<HypervisorSettingsResponse> {
	return bffFetch(BFFEndpoints.getHypervisorSettings, {
		method: 'GET',
		token
	});
}

export async function updateHypervisorSettings(
	payload: HypervisorSettingsPatch,
	token?: string
): Promise<HypervisorSettingsResponse> {
	return bffFetch(BFFEndpoints.updateHypervisorSettings, {
		method: 'PATCH',
		body: JSON.stringify(payload),
		token
	});
}

export async function applyHypervisorProfile(
	profileId: string,
	token?: string
): Promise<HypervisorSettingsResponse> {
	return bffFetch(`${BFFEndpoints.applyHypervisorProfile}/${profileId}`, {
		method: 'POST',
		token
	});
}

export async function listHypervisorProfiles(token?: string): Promise<{ profiles: HypervisorProfile[] }> {
	return bffFetch(BFFEndpoints.listHypervisorProfiles, {
		method: 'GET',
		token
	});
}

File: crates/chv-controlplane-store/src/hypervisor_settings.rs
use crate::{StoreError, StorePool};

#[derive(sqlx::FromRow, Debug, Clone)]
pub struct HypervisorSettingsRow {
    pub id: i64,
    pub cpu_nested: bool,
    pub cpu_amx: bool,
    pub cpu_kvm_hyperv: bool,
    pub memory_mergeable: bool,
    pub memory_hugepages: bool,
    pub memory_shared: bool,
    pub memory_prefault: bool,
    pub iommu: bool,
    pub rng_src: String,
    pub watchdog: bool,
    pub landlock_enable: bool,
    pub serial_mode: String,
    pub console_mode: String,
    pub pvpanic: bool,
    pub tpm_type: Option<String>,
    pub tpm_socket_path: Option<String>,
    pub profile_id: Option<String>,
    pub updated_at: String,
}

#[derive(Clone)]
pub struct HypervisorSettingsRepository {
    pool: StorePool,
}

impl HypervisorSettingsRepository {
    pub fn new(pool: StorePool) -> Self {
        Self { pool }
    }

    pub fn pool(&self) -> &StorePool {
        &self.pool
    }

    pub async fn get_settings(&self) -> Result<HypervisorSettingsRow, StoreError> {
        let row = sqlx::query_as::<_, HypervisorSettingsRow>(
            "SELECT * FROM hypervisor_settings WHERE id = 1",
        )
        .fetch_one(&self.pool)
        .await?;
        Ok(row)
    }
}
